`*.d.ts` files are not checked because of `skipLibCheck` in `tsconfig.json`

## Introduction

At some point, the `skipLibCheck` option was enabled in `tsconfig.json` because of mismatched types in some external `node_modules` dependencies (see [the comment](https://github.com/vaadin/web-components/pull/2254#issuecomment-883461112)).

However, this option doesn't prevent checking only `node_modules/**/*.d.ts`, but rather all the existing declaration files in the project, meaning that even the web components' declaration files don't get checked anymore.

That resulted in us missing some regressions in the type declarations for the web components. 
See the output of `yarn lint:types` with disabled option:

```
yarn run v1.22.10
$ tsc
packages/field-base/src/input-slot-mixin.d.ts:6:10 - error TS2724: '"./delegate-focus-mixin.js"' has no exported member named 'DelegatesFocusMixin'. Did you mean 'DelegateFocusMixin'?

6 import { DelegatesFocusMixin } from './delegate-focus-mixin.js';
           ~~~~~~~~~~~~~~~~~~~

packages/field-base/src/text-area-slot-mixin.d.ts:6:10 - error TS2724: '"./delegate-focus-mixin.js"' has no exported member named 'DelegatesFocusMixin'. Did you mean 'DelegateFocusMixin'?

6 import { DelegatesFocusMixin } from './delegate-focus-mixin.js';
           ~~~~~~~~~~~~~~~~~~~

packages/select/src/interfaces.d.ts:1:10 - error TS2305: Module '"./vaadin-select"' has no exported member 'SelectElement'.

1 import { SelectElement } from './vaadin-select';
           ~~~~~~~~~~~~~

packages/select/src/vaadin-select.d.ts:145:3 - error TS2416: Property 'helperText' in type 'Select' is not assignable to the same property in base type 'HTMLElement & ThemableMixin & ThemePropertyMixin & ElementMixin & DirMixin & SlotMixin & LabelMixin & FieldAriaMixin & DelegateFocusMixin'.
  Type 'string | null | undefined' is not assignable to type 'string'.
    Type 'undefined' is not assignable to type 'string'.

145   helperText: string | null | undefined;
      ~~~~~~~~~~

packages/vaadin-notification/src/vaadin-notification.d.ts:155:34 - error TS2304: Cannot find name 'TemplateResult'.

155   static show(contents: string | TemplateResult, options?: ShowOptions): NotificationElement;
                                     ~~~~~~~~~~~~~~

packages/vaadin-notification/src/vaadin-notification.d.ts:155:60 - error TS2304: Cannot find name 'ShowOptions'.

155   static show(contents: string | TemplateResult, options?: ShowOptions): NotificationElement;
                                                               ~~~~~~~~~~~
Found 6 errors.
```

## Proposal

As we can see above, there are no longer errors regarding `node_modules` dependencies so we can disable the `skipLibCheck` option in order to make the TypeScript compiler checking the project's declarations files again.

Disabling the option also supposes all the reported regressions to be fixed.

## Performance

Note, disabling the option is going to slow down the type checking.

**`skipLibCheck: true`**

```bash
time yarn lint:types

Executed in 1.91 secs
```

**`skipLibCheck: false`**

```bash
time yarn lint:types

Executed in 3.21 secs
```

**`skipDefaultLibCheck: true`**

```bash
time yarn lint:types

Executed in 2.95 secs
```

**Takeaway:** We could consider enabling `skipDefaultLibCheck` in order to reduce the cost of disabling `skipLibCheck`.

--- a/packages/field-base/src/helper-text-mixin.d.ts
+++ b/packages/field-base/src/helper-text-mixin.d.ts
@@ -18,7 +18,7 @@ interface HelperTextMixin extends SlotMixin {
   /**
    * String used for the helper text.
    */
-  helperText: string;
+  helperText: string | null | undefined;
 }
 
 export { HelperTextMixinConstructor, HelperTextMixin };

--- a/packages/field-base/src/input-field-mixin.js
+++ b/packages/field-base/src/input-field-mixin.js
@@ -113,7 +113,7 @@ const InputFieldMixinImplementation = (superclass) =>
     }
 
     /**
-     * Override an event listener from `DelegatesFocusMixin`.
+     * Override an event listener from `DelegateFocusMixin`.
      * @param {FocusEvent} event
      * @protected
      * @override
@@ -127,7 +127,7 @@ const InputFieldMixinImplementation = (superclass) =>
     }
 
     /**
-     * Override an event listener from `DelegatesFocusMixin`.
+     * Override an event listener from `DelegateFocusMixin`.
      * @param {FocusEvent} event
      * @protected
      * @override

--- a/packages/field-base/src/input-slot-mixin.d.ts
+++ b/packages/field-base/src/input-slot-mixin.d.ts
@@ -3,7 +3,7 @@
  * Copyright (c) 2021 Vaadin Ltd.
  * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
  */
-import { DelegatesFocusMixin } from './delegate-focus-mixin.js';
+import { DelegateFocusMixin } from './delegate-focus-mixin.js';
 import { SlotMixin } from './slot-mixin.js';
 import { InputMixin } from './input-mixin.js';
 
@@ -16,7 +16,7 @@ interface InputSlotMixinConstructor {
   new (...args: any[]): InputSlotMixin;
 }
 
-interface InputSlotMixin extends DelegatesFocusMixin, InputMixin, SlotMixin {
+interface InputSlotMixin extends DelegateFocusMixin, InputMixin, SlotMixin {
   /**
    * String used to define input type.
    */

--- a/packages/field-base/src/text-area-slot-mixin.d.ts
+++ b/packages/field-base/src/text-area-slot-mixin.d.ts
@@ -3,7 +3,7 @@
  * Copyright (c) 2021 Vaadin Ltd.
  * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
  */
-import { DelegatesFocusMixin } from './delegate-focus-mixin.js';
+import { DelegateFocusMixin } from './delegate-focus-mixin.js';
 import { SlotMixin } from './slot-mixin.js';
 import { InputMixin } from './input-mixin.js';
 
@@ -16,6 +16,6 @@ interface TextAreaSlotMixinConstructor {
   new (...args: any[]): TextAreaSlotMixin;
 }
 
-interface TextAreaSlotMixin extends DelegatesFocusMixin, InputMixin, SlotMixin {}
+interface TextAreaSlotMixin extends DelegateFocusMixin, InputMixin, SlotMixin {}
 
 export { TextAreaSlotMixinConstructor, TextAreaSlotMixin };

--- a/packages/select/src/interfaces.d.ts
+++ b/packages/select/src/interfaces.d.ts
@@ -1,4 +1,4 @@
-import { SelectElement } from './vaadin-select';
+import { Select } from './vaadin-select';
 
 /**
  * Function for rendering the content of the `<vaadin-select>`.
@@ -8,7 +8,7 @@ import { SelectElement } from './vaadin-select';
  *   DOM element. Append your content to it.
  * - `select` The reference to the `<vaadin-select>` element.
  */
-export type SelectRenderer = (root: HTMLElement, select?: SelectElement) => void;
+export type SelectRenderer = (root: HTMLElement, select?: Select) => void;
 
 /**
  * Fired when the `opened` property changes.

--- a/packages/select/src/vaadin-select.d.ts
+++ b/packages/select/src/vaadin-select.d.ts
@@ -140,11 +140,6 @@ declare class Select extends DelegateFocusMixin(
   placeholder: string | null | undefined;
 
   /**
-   * String used for the helper text.
-   */
-  helperText: string | null | undefined;
-
-  /**
    * When present, it specifies that the element is read-only.
    */
   readonly: boolean;

--- a/packages/vaadin-notification/src/vaadin-notification.d.ts
+++ b/packages/vaadin-notification/src/vaadin-notification.d.ts
@@ -1,4 +1,4 @@
-import { TemplateResult } from 'lit-html';
+import { TemplateResult } from 'lit';
 import { ElementMixin } from '@vaadin/vaadin-element-mixin/vaadin-element-mixin.js';
 import { ThemableMixin } from '@vaadin/vaadin-themable-mixin/vaadin-themable-mixin.js';
 import { ThemePropertyMixin } from '@vaadin/vaadin-themable-mixin/vaadin-theme-property-mixin.js';

--- a/packages/vaadin-notification/src/vaadin-notification.d.ts
+++ b/packages/vaadin-notification/src/vaadin-notification.d.ts
@@ -1,10 +1,8 @@
+import { TemplateResult } from 'lit-html';
 import { ElementMixin } from '@vaadin/vaadin-element-mixin/vaadin-element-mixin.js';
-
 import { ThemableMixin } from '@vaadin/vaadin-themable-mixin/vaadin-themable-mixin.js';
-
 import { ThemePropertyMixin } from '@vaadin/vaadin-themable-mixin/vaadin-theme-property-mixin.js';
-
-import { NotificationEventMap, NotificationPosition, NotificationRenderer } from './interfaces';
+import { NotificationEventMap, NotificationPosition, NotificationRenderer, ShowOptions } from './interfaces';
 
 /**
  * An element used internally by `<vaadin-notification>`. Not intended to be used separately.

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -5,8 +5,11 @@
     "moduleResolution": "node",
     "lib": [ "esnext", "es2018", "dom" ],
     "noEmit": true,
-    "skipLibCheck": true,
-    "strict": true
+    "strict": true,
+    // WARNING: Avoid enabling the `skipLibCheck` option as it disables type checking
+    // for the components' type declarations (e.g packages/button/src/vaadin-button.d.ts).
+    "skipLibCheck": false,
+    "skipDefaultLibCheck": true
   },
   "include": [
     "packages/**/test/typings"

